bart/sched/pelt.py

```diff
@@ -0,0 +1,377 @@
+        print "  u = {:5d}*(1 - y)    : {:6.3f}".format(Simulator._signal_max, self._geom_u)
+
+        self._df = None
+
+    def _geom_sum(self, u_1, active_us=0):
+        """
+        Geometric add the specified value to the u_1 series.
+
+        NOTE: the current implementation assume that the task was active
+              (active_us != 0) or sleeping (active_us == 0) a full PELT
+              sampling interval (by default 1024[us])
+        """
+        # Decay previous samples
+        u_1 *= self._geom_y
+        # Add current sample (if task was active)
+        if active_us:
+            u_1 += self._geom_u
+        return u_1
+
+    def stable_range(self, task):
+        """
+        Compute the stability ranges for the specified (periodic) task
+
+        Here we use:
+        - Decay factor:       y = 0.5^(1/half_life)
+        - Max stable value:       (1-y^r) / (1-y^p)
+        - Min stable value: y^i * (1-y^r) / (1-y^p)
+
+        :param task: the task we want to simulate the PELT signal for, by
+                     default is the task used to intialize the Simulator
+                     (when specified)
+        :type  task: Task
+
+        :return: A tuple representing the minimum and maximum value of the PELT
+                 singal once stable.
+
+        """
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
+
+        # Compute max value
+        max_util  = (1. - pow(self._geom_y, float(task.run_us)/Simulator._signal_max))
+        max_util /= (1. - pow(self._geom_y, float(task.period_us)/Simulator._signal_max))
+
+        # Compute min value, by decaying the maximum value
+        min_util  = (max_util * pow(self._geom_y, float(task.idle_us)/Simulator._signal_max))
+
+        return PELT_Range(Simulator._signal_max * min_util,
+                         Simulator._signal_max * max_util)
+
+
+    def signal(self, task, start_s=0, end_s=10):
+        """
+        Compute the PELT signal for the specified interval.
+
+        :param task: the task we want to simulate the PELT signal for, by
+                     default is the task used to intialize the Simulator
+                     (when specified)
+        :type  task: Task
+
+        :param start_s: the start time in [s]
+        :type  start_s: float
+
+        :param end_s: the end time in [s]
+        :type  end_s: float
+        """
+
+        # Computed PELT samples
+        samples = []
+        # Decay capping support
+        running_prev = True
+        capping = None
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
+
+        # Intervals (in PELT samples) for the signal to compute
+        start_us = int(1e6 * Simulator._sample_us/1e3 * start_s)
+        end_us   = int(1e6 * Simulator._sample_us/1e3 * end_s)
```

Comment:
These calculations are wrong. AFAICT it should be
```
# Intervals (in PELT samples) for the signal to compute
start_us = (start_s * 1e6) / sample_us
```
(in which case the variables should be renamed) or 
```
# Time (in us) for the signal to compute
start_us = start_s * 1e6
```
Currently the range of `.signal` is wrong:
```
In [9]: df = s.signal(t, end_s=1)

In [10]: df.index[-1]
Out[10]: 1.024
```
(only slightly wrong, because the sample period is roughly 1e3)